parlai/core/torch_ranker_agent.py

```diff
@@ -0,0 +1,351 @@
+#!/usr/bin/env python3
+
+# Copyright (c) 2017-present, Facebook, Inc.
+# All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree. An additional grant
+# of patent rights can be found in the PATENTS file in the same directory.
+
+import os
+from functools import lru_cache
+
+import torch
+from torch import nn
+
+from parlai.core.torch_agent import TorchAgent, Output
+from parlai.core.thread_utils import SharedTable
+from parlai.core.utils import round_sigfigs, padded_tensor
+
+class TorchRankerAgent(TorchAgent):
+    @staticmethod
+    def add_cmdline_args(argparser):
+        TorchAgent.add_cmdline_args(argparser)
+        agent = argparser.add_argument_group('TorchRankerAgent')
+        agent.add_argument(
+            '-tc', '--train_candidates', type=str, default='batch',
+            choices=['batch', 'inline', 'fixed', 'dict'],
+            help='the source of candidates during training')
+        agent.add_argument(
+            '-ec', '--eval_candidates', type=str, default='batch',
+            choices=['batch', 'inline', 'fixed', 'dict'],
+            help='the source of candidates during training')
+        agent.add_argument(
+            '-cands', '--fixed-candidates-path', type=str,
+            help='a text file of fixed candidates to use for all examples, '
+                'one candidate per line')
+
+    def __init__(self, opt, shared):
+        if shared:
+            super().__init__(opt, shared)
+            self.model = shared['model']
+            self.metrics = shared['metrics']
+        else:
+            # Must call _get_model_file first so that paths are updated if
+            # necessary (e.g., .dict file)
+            model_file = self._get_model_file(opt)
+            self.metrics = {'loss': 0.0, 'examples': 0, 'rank': 0}
+            opt['rank_candidates'] = True
+
+            super().__init__(opt, shared)
```

Comment:
should the call to `super().__init__` be outside the if? Seems weird it's called separately.

Reply:
Yeah, it's honestly a little weird to me too that super().__init__() gets sandwiched between other code. I think it should probably always be called first and then have things overwritten by the children as necessary. I ddi in this way to keep it compatible with past saved MemNN models, but if we're willing to break that (and I think it's worth it here), I can move some things up to TorchAgent so TorchRankerAgent isn't doing anything special for loading existing models (and I don't think it needs to).

Reply:
No, those models need to remain working; these pretrained models are among our most popular downloads.

Reply:
Ok, so I guess I'll leave this for now to not break things. But I think long-term it's worth being explicit somewhere about how we expect inheritance to work here. Is the TorchAgent just an abstract class meant to ensure certain methods exist in children, or is it going to be a base class that handles most non-model specific details? Because if the latter is the case, I think it would make sense to move model loading and initialization calls up to TorchAgent. (e.g., check if 'model_file' and/or 'dict_file' are in opt, and if they are, load them. if they're not, call build_model() and instantiate a new dictionary). But I get that modifying TorchAgent would touch a lot of things, so fine to not mess with it in this PR.

Reply:
MemNN had, basically:
```
model_file = self._get_model_file(opt)
super().__init__(opt, shared)
```
at the top of init, since we want to override the dict_file. it's fine if we do it on the shared instances too, even if a little wasted energy, to get it out of the if block.

We could move that to TorchAgent, seq2seq is doing the same thing.